design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,777 @@
+hyphen separated names, while system facing and aggregated roles use colon
+user-facing role. There is only one `crossplane:aggregate-to-admin` base role
+and only separated names. In the above example the base role is one-to-one with
+the one `crossplane-admin` user-facing role. In other cases one base role is
+shared by many user-facing roles - for example many user-facing namespace roles
+share one base role:
+
+```yaml
+---
+# The 'user-facing' edit role for the 'example' namespace.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane-ns-example-edit
+aggregationRule:
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-ns-edit: "true"
+      rbac.crossplane.io/base-of-ns-edit: "true"
+---
+# The 'base' role containing the fixed rules that always aggregate to all
+# namespace aligned edit cluster roles.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:aggregate-to-ns-edit
+  labels:
+    rbac.crossplane.io/aggregate-to-ns-edit: "true"
+    rbac.crossplane.io/base-of-ns-edit: "true"
+rules:
+- apiGroups: [""]
+  resources: [events]
+  verbs: [get, list, watch]
+- apiGroups: [""]
+  resources: [secrets]
+  verbs: ["*"]
+- apiGroups: [apiextensions.crossplane.io]
+  resources: ["*"]
+  verbs: [get, list, watch]
+```
+
+Note that aggregation may be transitive. For example the `crossplane-admin` role
+is identical to the `crossplane-edit` role except for a few extra rules in its
+base role. Provider and composite resource roles can therefore aggregate only to
+`crossplane-edit`, which in turn aggregates to `crossplane-admin`.
+
+### Composite Resource ClusterRole Mechanics
+
+When a composite resource is defined the RBAC manager will create two cluster
+roles. For example creating an XRD named `composites.example.org` would trigger
+the creation of the following roles:
+
+* `crossplane:composite:composites.example.org:aggregate-to-edit`
+* `crossplane:composite:composites.example.org:aggregate-to-view`
+
+The `aggregate-to-edit` role aggregates to the `crossplane`, `crossplane-admin`,
+and `crossplane-edit` roles, as well as any namespace-aligned `edit` role for a
+namespace in which the composite resource may claimed. The `aggregate-to-view`
+cluster role aggregates to the `crossplane-view` cluster role, and any `view`
+role for a namespace in which the composite resource may be claimed.
+
+Composite resources may be claimed in any namespace that has an annotation with
+the key `rbac.crossplane.io/composites.example.org`, and the value `xrd-enabled`
+(where `composites.example.org` is the name of the XRD). When the RBAC manager
+encounters a namespace with one or more such annotations, it creates `edit` and
+`view` cluster roles for that namespace.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: example
+  annotations:
+    rbac.crossplane.io/composites.example.org: xrd-enabled
```

Comment:
Nit: maybe something with `offer` owrd, like `xrd-offered`?

Reply:
I don't like using the term offered here - it conflates the act of creating a namespaced CRD with the act of managing RBAC roles that all that defined CR to be used in a namespace. "Offers a claim" is a platform builder concern, while "enables a claim in a namespace" is a platform operator concern.

Reply:
I could also go for something like `rbac.crossplane.io/composites.example.org: claim-accepted`. I feel like this would fit pretty well, i.e.:

1. Platform builder offers a claim for this XRD.
1. Platform operator accepts the claim in this namespace.